File: app/javascript/channels/chat_channel.js
import consumer from "./cable";

const createChatChannel = (gameId) => {
    return consumer.subscriptions.create(
        { channel: "ChatChannel", game_id: gameId },
        {
            connected() {
                console.log("Connected to chat channel for game", gameId);
            },
            disconnected() {
                console.log("Disconnected from chat channel");
            },
            received(data) {
                const messagesContainer = document.getElementById("messages");
                messagesContainer.insertAdjacentHTML("beforeend", data.message);
            },
            speak(message) {
                this.perform("speak", { message: message });
            },
        }
    );
};

export default createChatChannel;

File: app/channels/chat_channel.rb
class ChatChannel < ApplicationCable::Channel
  def subscribed
    stream_from "chat_#{params[:game_id]}"
  end

  def unsubscribed
    # Cleanup if necessary
  end

  def speak(data)
    message = Message.create!(
      content: data['message'],
      user: current_user,
      game_id: params[:game_id]
    )
    ActionCable.server.broadcast("chat_#{params[:game_id]}", message: render_message(message))
  end

  private

  def render_message(message)
    ApplicationController.renderer.render(partial: 'messages/message', locals: { message: message })
  end
end
